test/integration/template_database_test.go

```diff
@@ -2531,85 +2671,169 @@ func TestDatabaseStatefulSetVolumeSnapshot(t *testing.T) {
-			assert.Equal(t, "PersistentVolumeClaim", obj.Spec.VolumeClaimTemplates[0].Spec.DataSourceRef.Kind)
-			assert.Equal(t, "some-other-sm", obj.Spec.VolumeClaimTemplates[0].Spec.DataSourceRef.Name)
-			assert.Nil(t, obj.Spec.VolumeClaimTemplates[0].Spec.DataSourceRef.APIGroup)
-			assert.Nil(t, obj.Spec.VolumeClaimTemplates[0].Spec.DataSourceRef.Namespace)
+			archiveFound := false
+			journalFound := false
+			for _, volumeClaimTemplate := range obj.Spec.VolumeClaimTemplates {
+				if volumeClaimTemplate.Name == "archive-volume" || volumeClaimTemplate.Name == "journal-volume" {
+					if volumeClaimTemplate.Name == "archive-volume" {
+						archiveFound = true
+					} else {
+						journalFound = true
+					}
+					assert.Nil(t, volumeClaimTemplate.Spec.DataSourceRef.APIGroup)
+					assert.Equal(t, "PersistentVolumeClaim", volumeClaimTemplate.Spec.DataSourceRef.Kind)
+					assert.Equal(t, volumeClaimTemplate.Name+"-otherdb-0", volumeClaimTemplate.Spec.DataSourceRef.Name)
+					assert.Equal(t, "otherns", *volumeClaimTemplate.Spec.DataSourceRef.Namespace)
+				}
+			}
+			assert.True(t, archiveFound)
+			assert.True(t, journalFound)
 			assert.True(t, testlib.EnvContains(obj.Spec.Template.Spec.Containers[0].Env, "SNAPSHOT_RESTORED", "true"))
 			assert.True(t, testlib.EnvContains(obj.Spec.Template.Spec.Containers[0].Env, "BACKUP_ID", ""))
 		}
 	})
 
-	// Test starting with an old values.yaml file that does not have default values set for dataSourceRef
+	// Render volume snapshots with validation (enabled by default) and check that it fails due to missing volume snapshot
+	t.Run("testSnapshotRestoreValidation", func(t *testing.T) {
+		backupId := "123abc"
+		options := &helm.Options{
+			SetValues: map[string]string{
+				"database.snapshotRestore.backupId": backupId,
+			},
+		}
+
+		_, err := helm.RenderTemplateE(t, options, helmChartPath, "release-name", []string{"templates/statefulset.yaml"})
+		assert.NotNil(t, err)
+		assert.Contains(t, err.Error(), "Invalid data source: VolumeSnapshot/123abc-archive not found in namespace default")
+	})
+
+	// Render PVC data source with validation (enabled by default) and check that it fails due to missing PVC
+	t.Run("testDataSourceValidation", func(t *testing.T) {
+		options := &helm.Options{
+			SetValues: map[string]string{
+				"database.persistence.journalDataSource.kind":      "PersistentVolumeClaim",
+				"database.persistence.journalDataSource.name":      "journal-volume-otherdb-0",
+				"database.persistence.journalDataSource.namespace": "otherns",
+				"database.persistence.journalDataSource.apiGroup":  "null", // Unset the default
+				"database.sm.hotCopy.journalPath.enabled":          "true",
+			},
+		}
+
+		_, err := helm.RenderTemplateE(t, options, helmChartPath, "release-name", []string{"templates/statefulset.yaml"})
+		assert.NotNil(t, err)
+		assert.Contains(t, err.Error(), "Invalid data source: PersistentVolumeClaim/journal-volume-otherdb-0 not found in namespace otherns")
```

Comment:
I assume that we can't test the opposite scenario in the _template_ tests as we don't have a real K8s API server?

Reply:
Correct, there is no K8s cluster to test against.